snapcraft/internal/common.py

```diff
@@ -57,21 +58,73 @@ def assemble_env():
     return "\n".join(["export " + e for e in env])
 
 
+run_number: int = 0
+
+
 def _run(cmd: List[str], runner: Callable, **kwargs):
+    global run_number
+    run_number += 1
+
     assert isinstance(cmd, list), "run command must be a list"
+
+    lines: List[str] = list()
+
+    # Set shell.
+    lines.append("#!/bin/sh")
+
+    # Account for `env` parameter by populating exports.
+    # Ordering matters: assembled_env overrides `env` parameter.
+    cmd_env = kwargs.pop("env", None)
+    if cmd_env:
+        lines.append("#############################")
+        lines.append("# Exported via `env` parameter:")
```

Comment:
do we really want something like this? I was wishing on moving away from this assembled_env thing completely, seems this change would tie us more to "common".

Reply:
I personally found it quite useful for debugging rust stuff, and will certainly use it again in the future.  The idea is just to provide a tool for debugging commands without trying to guess or waste a lot of time trying to reproduce the environment.  OpenEmbedded/bitbake does a similar thing to capture each build step, and I always have found that incredibly useful in my workflow.

The next best idea I had would be to setup the environment for --debug to reproduce the environment of the failed command that executed.  This was perhaps easier to inspect. :D

When things change, like removing assembled_env, this can be updated, or even removed, of course.  But there's always going to be an environment, specific to a part, step, and command.  It should be easy to reproduce and inspect it.

Reply:
ok, we can keep it as long as we can remove assembled_env and you have a plan for that! :-)